libafl/src/mutators/numeric.rs

```diff
@@ -0,0 +1,491 @@
+            }
+
+            #[inline]
+            fn flip_bit_at(&mut self, offset: usize) {
+                *self ^= 1 << offset;
+            }
+
+            #[inline]
+            fn wrapping_inc(&mut self) {
+                *self = self.wrapping_add(1);
+            }
+
+            #[inline]
+            fn wrapping_dec(&mut self) {
+                *self = self.wrapping_sub(1);
+            }
+
+            #[inline]
+            fn twos_complement(&mut self) {
+                *self = self.wrapping_neg();
+            }
+
+            #[inline]
+            #[allow(trivial_numeric_casts, clippy::cast_possible_wrap)]
+            fn randomize<R: Rand>(&mut self, rand: &mut R) {
+                *self = rand.next() as $t;
+            }
+
+        }
+    )*)
+}
+
+impl_numeric_cast_randomize!( u8 u16 u32 u64 usize i8 i16 i32 i64 isize );
+
+// Macro to implement the Numeric trait for multiple integer types a u64 cannot be cast to
+macro_rules! impl_numeric_iterate_randomize {
+    ($($t:ty)*) => ($(
+        impl Numeric for $t {
+            #[inline]
+            fn flip_all_bits(&mut self) {
+                *self = !*self;
+            }
+
+            #[inline]
+            fn flip_bit_at(&mut self, offset: usize) {
+                *self ^= 1 << offset;
+            }
+
+            #[inline]
+            fn wrapping_inc(&mut self) {
+                *self = self.wrapping_add(1);
+            }
+
+            #[inline]
+            fn wrapping_dec(&mut self) {
+                *self = self.wrapping_sub(1);
+            }
+
+            #[inline]
+            fn twos_complement(&mut self) {
+                *self = self.wrapping_neg();
+            }
+
+            #[inline]
+            fn randomize<R: Rand>(&mut self, rand: &mut R) {
+                self.set_zero();
+
+                let byte_size = size_of::<$t>();
+                let bytes_per_rand = size_of::<u64>();
+
+                let mut current_rand = 0u64;
+
+                for byte_index in 0..byte_size {
+                    if byte_index % bytes_per_rand == 0 {
+                        current_rand = rand.next();
+                    }
+
+                    let rand_index = (byte_index % bytes_per_rand);
+                    let rand_byte = ((current_rand >> (8 * rand_index)) & 0xFF) as u8;
+                    *self |= Self::from(rand_byte) << (8 * byte_index);
+                }
+            }
+
+        }
+    )*)
+}
+
+// Apply the macro to all desired integer types
+impl_numeric_iterate_randomize! { u128 i128 }
+
+/// Bitflip mutation for integer-like inputs
+#[derive(Debug)]
+pub struct BitFlipMutator;
+
+impl<I, S> Mutator<I, S> for BitFlipMutator
+where
+    S: HasRand,
+    I: Numeric,
+{
+    fn mutate(&mut self, state: &mut S, input: &mut I) -> Result<MutationResult, Error> {
+        let offset = state.rand_mut().choose(0..size_of::<I>()).unwrap();
+        input.flip_bit_at(offset);
+        Ok(MutationResult::Mutated)
```

Comment:
To avoid superfluous executions, consider requiring `Numeric: Copy` and
```suggestion
        let original = *input;
        input.flip_bit_at(offset);
        if original == *input {
            Ok(MutationResult::Skipped)
        } else {
            Ok(MutationResult::Mutated)
        }
```
and similarly for the other mutations